### DISCUSSION TITLE
How to catch a terminal output of "fyne.io/fyne/v2/dialog"

### DISCUSSION THREAD

**Discussion by jwill0182** [fyne-io/fyne#3614] [Q&A] [answered]
Hello,

I´d like to catch the terminal output / errors of "fyne.io/fyne/v2/dialog" and pump them into a log file. 

**Terminal output:**
2023/01/30 10:53:42 Fyne error:  Getting favorite locations
2023/01/30 10:53:42   Cause: uri is not listable
2023/01/30 10:53:42   At: /go/pkg/mod/fyne.io/fyne/v2@v2.2.4/dialog/file.go:277
2023/01/30 10:53:54 Fyne error:  Error getting file attributes
2023/01/30 10:53:54   Cause: The process cannot access the file because it is being used by another process.
2023/01/30 10:53:54   At: /go/pkg/mod/fyne.io/fyne/v2@v2.2.4/dialog/file_windows.go:78
2023/01/30 10:53:54 Fyne error:  Error getting file attributes
2023/01/30 10:53:54   Cause: The process cannot access the file because it is being used by another process.
2023/01/30 10:53:54   At: /go/pkg/mod/fyne.io/fyne/v2@v2.2.4/dialog/file_windows.go:78
2023/01/30 10:53:54 Fyne error:  Error getting file attributes
2023/01/30 10:53:54   Cause: The process cannot access the file because it is being used by another process.
2023/01/30 10:53:54   At: /go/pkg/mod/fyne.io/fyne/v2@v2.2.4/dialog/file_windows.go:78
2023/01/30 10:53:54 Fyne error:  Error getting file attributes
2023/01/30 10:53:54   Cause: The process cannot access the file because it is being used by another process.
2023/01/30 10:53:54   At: /go/pkg/mod/fyne.io/fyne/v2@v2.2.4/dialog/file_windows.go:78

In the example below the terminal output of fmt.println is written to pipe, readable and can be sent to a logfile - that works perfectly well. Now, I´d like to achieve the same result with a fyne dialog - Does anybody can please give me any hint / idea on how to do this?

`old := os.Stdout
r, w, _ := os.Pipe()
os.Stdout = w
os.Stderr = w

fmt.Println("Hello, playground") // this gets captured
file_dialog = dialog.NewFolderOpen(func(uri fyne.ListableURI, err error) {}, pMyWindow) //this doesn´t get captured
w.Close()

out, _ := ioutil.ReadAll(r)
os.Stdout = old

tmp := string(out[:])

M.Log.Println("Captured:" + tmp) // prints: Captured: Hello, playground`

**Comment by Jacalz** [comment, marked as answer]
> file_dialog = dialog.NewFolderOpen(func(uri fyne.ListableURI, err error) {}, pMyWindow) //this doesn´t get captured

What do you mean with this? It does not print anything to the terminal; you are just creating a folder dialog object. What are you expecting to happen?

If you want to print the object that you opened, then you need to put a println call within the function you pass to the dialog.

**Reply by jwill0182** [reply]
If the dialog is opened e.g. an error is written to the terminal which is caught in loadFavorites() funciton of file.go part of `"fyne.io/fyne/v2/dialog"` 

`func (f *fileDialog) loadFavorites() {
	favoriteLocations, err := getFavoriteLocations()
	if err != nil {
		fyne.LogError("Getting favorite locations", err)
	}`

I like to pipe this error into a log file together with the other above mentioend ones.

**Reply by Jacalz** [reply]
The `fyne.LogError` call just prints using the `log` library. You might be able to use https://pkg.go.dev/log#SetOutput to set an output writer.

**Reply by jwill0182** [reply]
Maybe this is a misunderstanding. I´ve already created a custom logger which sets the output location via method param. This works perfectly well hence terminal outputs e.g. fmt.Println outputs are sent to an logfile.

`func InitBaseLogging(sPwd string) {
	pFile, err := os.OpenFile(sPwd, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0666)
	if err != nil {
		log.Fatal(err)
	}
	Log = log.New(pFile, "BASIC: ", log.Ldate|log.Ltime|log.Lshortfile)
}`

I´m really looking for the way to get the output of `Fyne.LogError`  in order to redirect it a logfile or redirect the output /dev/null 2>&1 at least.

**Reply by Jacalz** [reply]
In that case I am not quite sure that I understand what you are asking about. My comment should, using `log.SetOutput()`, redirect all the output of `fyne.LogError` to a writer, which of course can be any file (or any other writer you prefer). In what way is this not the desired effect?

**Reply by jwill0182** [reply]
It works now 👍. Thx.

`	pFile, err := os.OpenFile(sPwd, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0666)
	if err != nil {
		log.Fatal(err)
	}

	Log = log.New(pFile, "BASIC: ", log.Ldate|log.Ltime|log.Lshortfile)
	InfoLogger = log.New(pFile, "INFO: ", log.Ldate|log.Ltime|log.Lshortfile)
	WarningLogger = log.New(pFile, "WARNING: ", log.Ldate|log.Ltime|log.Lshortfile)
	ErrorLogger = log.New(pFile, "ERROR: ", log.Ldate|log.Ltime|log.Lshortfile)

	log.SetOutput(pFile)`

**Reply by Jacalz** [reply]
Great. Feel free to mark this thread as the solution then :)
